### PR TITLE
Fix: RandomState.choice reproducibility

### LINKED ISSUE #738: RandomState.choice uses global random state
Current master (77c0ac33ada105bc8e97065a5ae4041fea7d5b13)

### Code
```
import cupy

rs = cupy.random.RandomState()

rs.seed(100)
print(rs.uniform())

rs.seed(100)
print(rs.uniform())

rs.seed(100)
print(rs.choice(3, size=10))

rs.seed(100)
print(rs.choice(3, size=10))
```

### Output
```
0.6116323343956134
0.6116323343956134
[0 2 2 2 0 0 1 2 0 1]
[0 0 0 2 0 0 1 0 2 1]
```

### PR DESCRIPTION
**Pull request by niboshi**
Fixes #738 

I had to refactor unit test, so that reproducibility of generators is checked every time generating method is called.

### CODE REVIEW

File: tests/cupy_tests/random_tests/test_generator.py [hunk 1]
```diff
@@ -208,99 +256,99 @@ def test_randn_invalid_argument(self):
...
         for val in vals:
             self.assertEqual(val.shape, (2, 3))
         self.assertEqual(min(_.min() for _ in vals), 0)
         self.assertEqual(max(_.max() for _ in vals), 10)
 
     @condition.repeat(3, 10)
     def test_bound_2(self):
-        vals = [self.rs.interval(2, None).get() for _ in range(20)]
+        vals = self.generate_many(2, None, _count=20)
+        vals = [_.get() for _ in vals]
```
Comment (unnonouno) [inline tests/cupy_tests/random_tests/test_generator.py]: Stop using `_`. How about like `val`?

File: tests/cupy_tests/random_tests/test_generator.py [hunk 2]
```diff
@@ -208,99 +256,99 @@ def test_randn_invalid_argument(self):
...
         self.assertEqual(max(vals), 2)
 
     @condition.repeat(3, 10)
     def test_goodness_of_fit(self):
         mx = 5
         trial = 100
-        vals = [self.rs.interval(mx, None).get()
-                for _ in six.moves.xrange(trial)]
+        vals = self.generate_many(mx, None, _count=trial)
+        vals = [_.get() for _ in vals]
```
Comment (unnonouno) [inline tests/cupy_tests/random_tests/test_generator.py]: ditto

File: tests/cupy_tests/random_tests/test_generator.py [hunk 3]
```diff
@@ -337,8 +383,9 @@ def test_dtype_shape(self):
 
     @condition.repeat(3, 10)
     def test_bound(self):
-        vals = [self.rs.choice(a=self.a, size=self.size, p=self.p).get()
-                for _ in range(20)]
+        vals = self.generate_many(
+            a=self.a, size=self.size, p=self.p, _count=20)
+        vals = [_.get() for _ in vals]
```
Comment (unnonouno) [inline tests/cupy_tests/random_tests/test_generator.py]: ditto

File: tests/cupy_tests/random_tests/test_generator.py [hunk 4]
```diff
@@ -372,8 +417,9 @@ def test_dtype_shape(self):
...
```
Comment (unnonouno) [inline tests/cupy_tests/random_tests/test_generator.py]: dito

File: tests/cupy_tests/random_tests/test_generator.py [hunk 5]
```diff
@@ -383,24 +429,22 @@ def test_bound(self):
...
-        self.rs.seed(testing.generate_seed())
+    target_method = 'choice'
 
     @condition.repeat(3, 10)
     def test_goodness_of_fit(self):
         trial = 100
-        vals = [self.rs.choice(3, 1, True, [0.3, 0.3, 0.4]).get()
-                for _ in six.moves.xrange(trial)]
+        vals = self.generate_many(3, 1, True, [0.3, 0.3, 0.4], _count=trial)
+        vals = [_.get() for _ in vals]
```
Comment (unnonouno) [inline tests/cupy_tests/random_tests/test_generator.py]: ditto

File: tests/cupy_tests/random_tests/test_generator.py [hunk 6]
```diff
@@ -208,99 +256,99 @@ def test_randn_invalid_argument(self):
...
 
-        v = self.rs.interval(0xffffffff, 2)
+        v = self.generate(0xffffffff, 2)
         self.assertEqual(v.dtype, numpy.uint32)
 
     @condition.repeat(3, 10)
     def test_bound_1(self):
-        vals = [self.rs.interval(10, (2, 3)).get() for _ in range(10)]
+        vals = self.generate_many(10, (2, 3), _count=10)
+        vals = [_.get() for _ in vals]
```
Comment (unnonouno) [inline tests/cupy_tests/random_tests/test_generator.py]: ditto

File: cupy/random/generator.py [hunk 7]
```diff
@@ -14,6 +14,20 @@
 from cupy.cuda import curand
 
 
+_gumbel_kernel = None
+
+
+def _get_gumbel_kernel():
+    global _gumbel_kernel
```
[the remark this answers is not in the document]
Comment (niboshi) [reply cupy/random/generator.py]: Yes, but `ElementwiseKernel.__init__` includes some cache lookups and object handling, which is not free. That can be saved with this.
The original code caches it as well and it's consistent with other part, such as `cupy.math.sumprod`.

The intent of wrapping with a function is to reduce the time taken in `import cupy`.

**Comment by niboshi** [comment]
Thank you. Fixed `_` usage.

File: cupy/random/generator.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (unnonouno) [reply cupy/random/generator.py]: ok

**Comment by unnonouno** [comment]
LGTM

**Pull request merged**